src/Microsoft.AspNet.Razor/Parser/HtmlMarkupParser.Document.cs

```diff
@@ -42,58 +43,102 @@ private void ScanTagInDocumentContext()
             {
                 if (NextIs(HtmlSymbolType.Bang))
                 {
-                    AcceptAndMoveNext(); // Accept '<'
-                    BangTag();
+                    // Checking to see if we meet the conditions of a special '!' tag: <!DOCTYPE, <![CDATA[, <!--.
+                    if (!AtBangEscape(lookahead: 1))
+                    {
+                        AcceptAndMoveNext(); // Accept '<'
+                        BangTag();
+                        return;
+                    }
+
+                    // We should behave like a normal tag that has a parser escape, fall through to the normal 
+                    // tag logic.
                 }
                 else if (NextIs(HtmlSymbolType.QuestionMark))
                 {
                     AcceptAndMoveNext(); // Accept '<'
                     XmlPI();
+                    return;
                 }
-                else
-                {
-                    Output(SpanKind.Markup);
 
-                    // Start tag block
-                    var tagBlock = Context.StartBlock(BlockType.Tag);
+                Output(SpanKind.Markup);
 
-                    AcceptAndMoveNext(); // Accept '<'
+                // Start tag block
+                var tagBlock = Context.StartBlock(BlockType.Tag);
 
-                    if (!At(HtmlSymbolType.ForwardSlash))
-                    {
-                        // Parsing a start tag
-                        var scriptTag = At(HtmlSymbolType.Text) &&
-                                        string.Equals(CurrentSymbol.Content, "script", StringComparison.OrdinalIgnoreCase);
-                        Optional(HtmlSymbolType.Text);
-                        TagContent(); // Parse the tag, don't care about the content
-                        Optional(HtmlSymbolType.ForwardSlash);
-                        Optional(HtmlSymbolType.CloseAngle);
-
-                        if (scriptTag)
-                        {
-                            Output(SpanKind.Markup);
-                            tagBlock.Dispose();
-
-                            SkipToEndScriptAndParseCode();
-                            return;
-                        }
-                    }
-                    else
+                AcceptAndMoveNext(); // Accept '<'
+
+                if (!At(HtmlSymbolType.ForwardSlash))
+                {
+                    OptionalBangEscape();
+
+                    // Parsing a start tag
+                    var scriptTag = At(HtmlSymbolType.Text) &&
+                                    string.Equals(CurrentSymbol.Content, "script", StringComparison.OrdinalIgnoreCase);
+                    Optional(HtmlSymbolType.Text);
+                    TagContent(); // Parse the tag, don't care about the content
+                    Optional(HtmlSymbolType.ForwardSlash);
+                    Optional(HtmlSymbolType.CloseAngle);
```

Comment:
We need to be able to handle malformed tags. Therefore the only thing not optional in a tag is the beginning `<` and tag name.